Implement Directus Auth in Next.js 14

emmanuelhashy opened the pull request:
Close #124

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,1105 @@
+**Setting Public Role Permission for Directus Users**
+Now, you should find two categories of users when you navigate to the User Directory: Administrator and Author.
+To ensure that a new user on your frontend application can only sign up as an author, navigate to the Access Control settings, click on the Author role, look at the info icon in the top right corner of the screen, and copy the `Primary Key`.
+
+![Directus author role settings](directus-role-settings.png)
+
+Navigate to the Public role settings and in the `directus_users` system collection, configure custom create permissions to allow user creation where `role` equals the Author role `Primary Key` you copied earlier.
+
+![Directus public role field validation settings](public-role-field-validation-settings.png)
+
+Click on `Field Permission` and check the fields required to create a user.
+
+![Directus public role field permission settings](public-role-field-permission-settings.png)
+
+The required fields are:
+
+- first_name
+- last_name
+- email
+- password
+- role
+
+**Creating posts collection**
+Navigate to `Settings/Data Model` and create a new collection named `posts`. This is where all the posts created by users with an Author role will be stored.
+
+![Directus posts collection setup](Directus-posts-collection-setup.png)
+
+Check the fields that you want returned on the `posts` data
+
+![Directus posts optional fields settings](Directus-posts-optional-fields-settings.png)
+
+Create additional fields for the post:
+
+- title => Input field
+- content => WYSIWYG
+- image => Image
+
+The `title` and `content` fields should be required.
+
+![Directus post additional fields settings](![alt text](Directus-post-additional-fields-settings-1.png)
```
Could you structure this section the same way as our iOS Auth post? 

```
## Allowing Public Role To Create Users

...

## Creating a New Role
```

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,1105 @@
+  avatar_file_name: 'profile-photo.jpeg'
+---
+
+## Introduction
+Next.js is a React.js server-side rendering framework that allows front-end developers to build full-stack applications with a single codebase. It has turborpack the rust-based bundler built-in for enhanced code optimization and fast code compilation. It’s a great choice for performance, efficiency and developer experience. In this tutorial, you will learn how to implement Authentication on a Nex ...
```
```suggestion
## Introduction
In this tutorial, you will learn how to integrate a Next.js project with Directus Auth. You'll cover registering, logging in, logging out, viewing all posts from all users, creating a post, and editing and deleting posts from your user account.
```

If they hit this page, they know what Next.js is - no need to explain.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,1105 @@
+To follow along with this tutorial, you will need the following knowledge and tools:
+
+- [Node.js](https://nodejs.org/en/download/) v18 and above installed on your computer.
+- A Directus project - you can use [Directus Cloud or run it yourself](https://docs.directus.io/getting-started/quickstart).
```
```suggestion
- A Directus project - follow our [quickstart guide](https://docs.directus.io/blog/docs.directus.io/getting-started/quickstart.html) if you don't already have one.
```

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,1105 @@
+
+- [Node.js](https://nodejs.org/en/download/) v18 and above installed on your computer.
+- A Directus project - you can use [Directus Cloud or run it yourself](https://docs.directus.io/getting-started/quickstart).
+- Prior experience with Nextjs.
```
```suggestion
- Prior experience with Next.js.
```

Can you go through the whole article and make sure you're properly stylizing "Next.js" please?

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,1105 @@
+- [Node.js](https://nodejs.org/en/download/) v18 and above installed on your computer.
+- A Directus project - you can use [Directus Cloud or run it yourself](https://docs.directus.io/getting-started/quickstart).
+- Prior experience with Nextjs.
+- Clone the [project repo](https://github.com/emmanuelhashy/Next-directus-auth-demo) to get the CSS styles for the demo project.
```
Remove this - this project is meant to be from scratch. Anything you've done manually that needs to be done by the reader needs to be explained right after this list.

phazonoverload replied:
Please also remove all styling configuration - 100% of it - from this project. It's not important.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,1105 @@
+Would you like to use `src/` directory? Yes
+Would you like to use App Router? Yes
+Would you like to customize the default import alias (@/*)? Yes
+```
+
+After the prompts, `create-next-app` will create a folder with your project name and install the required dependencies.
+
+Install the required dependencies:
+
+```bash
+    npm i next-auth @directus/sdk
+```
+
+## Set Up the ENV Variables
+
+Create a `.env` file in your Next.js project root directory and add the following:
+
+```text
+NEXTAUTH_URL=http://localhost:3000
+NEXTAUTH_SECRET=YOUR_NEXT_RANDOM_SECRET
+USER_ROLE=USER_ROLE_ID_FROM_YOUR_DIRECTUS_PROJECT
+BACKEND_URL=YOUR_DIRECTUS_PROJECT_URL
+NEXT_PUBLIC_URL=http://localhost:3000
+```
+
+You can access your Directus project from the Next.js application via `REST`, `GraphQL` and `SDK`. 
+
+## Set Up the Directus SDK
+
+Create an instance of the Directus SDK that will be used across the next.js application.
+In the `next-blog` project `src` directory, create a new directory called `lib` and a new file called `directus.js` inside `lib`, then add the following code:
+
+```javascript
+import { createDirectus, rest, authentication } from '@directus/sdk';
+const directus = createDirectus(process.env.BACKEND_URL)
+  .with(authentication("cookie", {credentials: "include", autoRefresh: true}))
+export default directus;
+```
+
+This code snippet sets up a client for interacting with the Directus backend with cookie-based authentication, ensuring that credentials are included in requests and enabling automatic token refresh.
````
Remove all of this in favor of linking to this document: https://docs.directus.io/guides/headless-cms/build-static-website/next

This post is meant to just be about auth. This has the added benefit of having some collections already set up. Make this a requirement in the **Before You Start**.

Obviously, you won't be able to add the `USER_ROLE` env var yet, but that's ok. Make that a step at the end of the "Creating a New Role" section.

phazonoverload reviewed (changes requested):
I've started to give a review but want to pause for some overall feedback. I think you're spending a lot of time in this post boilerplating a project "correctly" with things like the navigation. I want to strip most of that away and just get to the core auth. -> register, login, protecting routes, validating auth for create/update/delete operations. 

I also didn't expect you to use NextAuth.js in this project as we already have a post on this topic. I expected this to be free from auth libraries. Is that viable?

emmanuelhashy commented:
> I also didn't expect you to use NextAuth.js in this project as we already have a post on this topic. I expected this to be free from auth libraries. Is that viable?

@phazonoverload Absolutely, I understand your perspective. Developing a custom authentication solution will be time-consuming. I decided to use NextAuth.js because it is widely used in the Next.js ecosystem and provides built-in security features. 

Also, I have addressed all the comments. Feel free to let me know if you need anything else.

phazonoverload commented:
We already have a post using NextAuth and that isn't what was in your original proposal and not what we accepted. Will you be able to write this without an auth library? If not, then unfortunately this post isn't one we're able to publish given we have one already on this very topic.

emmanuelhashy commented:
@phazonoverload Alright, I will write the post without an auth library.

emmanuelhashy commented:
@phazonoverload I have rewritten the draft without an auth library.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+- Prior experience with Next.js.
+
+<!-- ## Your Sections Here -->
+## Allowing Public Role To Create Users
+
+1. Navigate to `Settings`.
+2. Select `Access Control`.
+3. Click on the `Public Role`.
+4. Access the `System Collection`.
+5. Locate `directus_users`.
+6. Under the `Create` option, choose `Use Custom`.
+7. Enable field permissions for `first_name`, `last_name`, `role`, `Email` and `Password` options.
+![Directus public role field permission settings](public-role-field-permission-settings.png)
+8. Enable field validation where `role` equals the Author role `Primary Key`
+![Directus public role field validation settings](public-role-field-validation-settings.png)
```
```suggestion
## Allowing Public Registration

When a user first registers, they will be unauthenticated. In Directus, the permissions of unauthenticated users are controlled by the public role.

In your access control settings open the public role. Find `directus_users` under system collections, and then set custom permissions.

1. In field permissions, enable only `first_name`, `last_name`, `email` and `password` options. If you want users to provide other data at the time of registration, enable the respective field.
2. In field validation, require `first_name`, `last_name`, `email`, and `password` to not be empty.
3. In field presets, set the `role` to the `id` of the author role.

This combination of settings in a custom permission will ensure users provide the required fields, and automatically set the role.
```

Your approach is valid, but it's safer to not expose the role setting to the client. By not allowing the field to be set by the public role, and using a field preset, new users are given the role automatically. 

This section also has to go after the creation of the author role, as it's referenced here.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+8. Enable field validation where `role` equals the Author role `Primary Key`
+![Directus public role field validation settings](public-role-field-validation-settings.png)
+
+## Creating posts collection
+
+1. Navigate to `Settings`.
+2. Select `Data Model`.
+3. Create a new collection named `posts`.
+4. Check the fields that you want returned on the `posts` data
+![Directus posts optional fields settings](Directus-posts-optional-fields-settings.png)
+5. Create additional fields for the post:
+
+- title => Input field
+- content => WYSIWYG
+- image => Image
+The `title` and `content` fields should be required.
+![Directus post additional fields settings](Directus-post-additional-fields-settings-1.png)
```
```suggestion
## Creating a Posts Collection

Create a new `posts` collection. Ensure `user_created` is created as an optional field as it will be used to lock down permissions so a user can only edit or delete their own posts. Create the following additional fields:

- `title`: a text input field
- `content`: a WYSIWYG field
- `image`: an image field

- title => Input field
- content => WYSIWYG
- image => Image
```

Required fields, I think, are not strictly required by this project so lets drop the instruction.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,995 @@
+The `title` and `content` fields should be required.
+![Directus post additional fields settings](Directus-post-additional-fields-settings-1.png)
+
+## Creating a New User Role
+
+1. On the Access Control page, click the plus button.
+2. Name your new role, as `Author`.
+3. Under the "Posts" collection, enable `Create` and `Read`.
+4. For `Edit` and `Delete`, select `Use Custom`.
+5. Add Filter `user_created -> id Equals $CURRENT_USER.id` for both update and delete options.
+6. This configuration ensures that users can read all posts but are restricted to updating and deleting only their own posts.
+7. Retrieve the `Primary Key` associated with the `Author` role by selecting the `Author Role` option and then tapping the information icon where you'll be able to copy the `Primary Key`
+8. On the Access Control page, click `Public`. Click `create` for `directus_users` from the options, then click on `Use custom`. Proceed to `field presets` and paste the Primary Key.
+
+By following these steps users created by the public role will be given the `Author` role.
+
+Add the following to your `.env` file:
+
+```text
+USER_ROLE=USER_ROLE_ID_FROM_YOUR_DIRECTUS_PROJECT
+```
````
````suggestion
## Creating a New User Role

Create a new role called `author`. Enable create and read permissions for the `posts` collection. For update and delete permissions, use custom rules and apply the following filter:

```
user_created -> id Equals $CURRENT_USER.id
```

This configuration ensures that users can read all posts but are restricted to updating and deleting only their own posts.

Retrieve the `Primary Key` associated with the `Author` role by selecting the `Author Role` option and then tapping the information icon where you'll be able to copy the `Primary Key`
````

As a field preset is being used, you do not need to expose the role to the client.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,995 @@
+```
+
+## Setting up Layout Configuration
+
````
Can this be removed / minimized and just straight into registration? It feels like a lot of boilerplate.

Perhaps start with registration, then login, then altering the UI based on state (the visible nav links), and then protecting routes. Read, create, update, and, finally, delete.

emmanuelhashy replied:
@phazonoverload This is an essential piece of the Next.js App. Removing it will result in an error after a successful login. I'll minimize it instead.

phazonoverload replied:
Minimizing is fine :)

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+        last_name,
+        email,
+        password,
+        role: process.env.USER_ROLE,
```
Remove this - not needed. Also the import.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,995 @@
+};
+```
+
+These functions handle user logout, registration, login and users session using the Directus SDK.
````
Can you spend a sentence explaining the userSession function a bit more

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+
+This will ensure that only authenticated users can access the create post and update post pages.
+
+## Implementing Server Actions for Auth Components
+
+Server Actions are **asynchronous functions** that are executed on the server. They can be used in Server and Client Components to handle form submissions and data mutations in Next.js applications.
+
+To send user credentials to the `Directus` backend on the user register form submission, navigate to the `lib` directory, create `action.js` file and add the following:
```
Can you please start this as a minimal file. And then add functions as we get to their respective sections?

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,995 @@
+```javascript
+"use server";
+import { revalidatePath } from "next/cache";
+import { directusLogin } from "./auth";
````
not mentioned elsewhere?

emmanuelhashy replied:
@phazonoverload `directusLogin` is used inside the `login` function in action.js while `revalidatePath` is used in the `addPost` function in action.js.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+  uploadFiles,
+  withToken,
+} from "@directus/sdk";
+import directus from "@/lib/directus";
```
not mentioned elsewhere?

emmanuelhashy replied:
@phazonoverload  `withToken` is used inside the `getAuthUser` function in action.js. I have removed `uploadFiles` as it's no longer in use.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+
+These functions handle user logout, registration, login and users session using the Directus SDK.
+
+## Creating the Auth Pages
```
This whole section is a bit jarring in how it's structured. Can we remove this H2 and have two H2s instead: "Implementing Registration" and "Implementing Login". Under each, see each to it's conclusion. It doesn't need to have subheadings, but structurally I would something like this

- Server action
- Form
- Page
- Test

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
````diff
@@ -0,0 +1,995 @@
+};
+```
+
+The `getPosts` function makes a request to Directus using the `directus.request()` method, specifying to read items from the "posts" collection. Within the request, it specifies the fields to retrieve for each post. The asterisk signifies all fields under its associated field.
````
Remove - this is a safe assumption of existing knowledge.

phazonoverload commented on implement-directus-auth-in-nextjs-14/index.md:
```diff
@@ -0,0 +1,995 @@
+          </form>
+          </div>}
+        </div>
+        <div dangerouslySetInnerHTML={{ __html: post.content }}></div>
```
Please make note that this can only be used for trusted content.

phazonoverload reviewed (changes requested):
Good! A couple of notes: 

- You reference some files that don't exist. I've left a couple of comments where this is the case - I assume this is a hangover of moving from using NextAuth to a direct implementation. Please do a  quick check of the post for any other examples of this. 
- Structurally I've left some comments - let's focus on seeing an entire piece of functionality through from start to end with registration and login. This will be much easier to follow and have a quicker 'win' for the reader, which is always good.
- How do you feel about removing image upload from this post? That's somewhat out of the scope of auth which is the main part of this post and only makes it longer. 

Otherwise, very happy and I reckon this will only need one more minor review once you've implemented these changes. 

My other other request is to crop all of the screenshots to remove whitespace and browser UI.

emmanuelhashy commented:
@phazonoverload I have made the requested changes. You can have a look.

phazonoverload commented:
You've not addressed all of my requested changes: 

https://github.com/directus-labs/guest-authoring/pull/152#discussion_r1591938709

https://github.com/directus-labs/guest-authoring/pull/152#discussion_r1591938896

Can you please do this, double check everything works, and then @ me. I will review this expediently once you have.

phazonoverload commented:
Thank you for writing this, and we are happy to consider it done in it's current state. My team will schedule it in to be published. 

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so we'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work.

Could you email me at [devrel@directus.io](mailto:devrel@directus.io) and we'll share what we need to get you paid. Thanks!
